src/change_stream.ts

```diff
@@ -860,127 +861,74 @@ export class ChangeStream<
     // between resumeToken and startAtOperationTime if we need to reconnect the cursor
     this.options.startAtOperationTime = undefined;
 
-    // Return the change
-    if (!callback) return this.emit(ChangeStream.CHANGE, change);
-    return callback(undefined, change);
+    return change;
   }
 
   /** @internal */
-  private _processError(error: AnyError, callback?: Callback) {
-    const cursor = this.cursor;
-
+  private _processErrorStreamMode(error: AnyError) {
     // If the change stream has been closed explicitly, do not process error.
-    if (this[kClosed]) {
-      // TODO(NODE-3485): Replace with MongoChangeStreamClosedError
-      if (callback) callback(new MongoAPIError(CHANGESTREAM_CLOSED_ERROR));
-      return;
-    }
-
-    // if the resume succeeds, continue with the new cursor
-    const resumeWithCursor = (newCursor: ChangeStreamCursor<TSchema, TChange>) => {
-      this.cursor = newCursor;
-      this._processResumeQueue();
-    };
-
-    if (cursor && isResumableError(error, maxWireVersion(cursor.server))) {
-      this.cursor = undefined;
+    if (this[kClosed]) return;
 
-      // stop listening to all events from old cursor
+    if (isResumableError(error, this.cursor.maxWireVersion)) {
       this._endStream();
-
-      // close internal cursor, ignore errors
-      cursor.close();
+      this.cursor.close();
 
       const topology = getTopology(this.parent);
-      this._waitForTopologyConnected(topology, { readPreference: cursor.readPreference }, err => {
-        // if the topology can't reconnect, close the stream
-        if (err) return this._closeWithError(err, callback);
-
-        // create a new cursor, preserving the old cursor's options
-        const newCursor = this._createChangeStreamCursor(cursor.resumeOptions);
-
-        // attempt to continue in emitter mode
-        if (!callback) return resumeWithCursor(newCursor);
-
-        // attempt to continue in iterator mode
-        newCursor.hasNext(err => {
-          // if there's an error immediately after resuming, close the stream
-          if (err) return this._closeWithError(err);
-          resumeWithCursor(newCursor);
-        });
+      topology.selectServer(this.cursor.readPreference, {}, err => {
+        if (err) return this._closeEmitterModeWithError(err);
+        this.cursor = this._createChangeStreamCursor(this.cursor.resumeOptions);
       });
-      return;
+    } else {
+      this._closeEmitterModeWithError(error);
     }
-
-    // if initial error wasn't resumable, raise an error and close the change stream
-    return this._closeWithError(error, callback);
   }
 
   /** @internal */
-  private _getCursor(callback: Callback<ChangeStreamCursor<TSchema, TChange>>) {
+  private _processErrorIteratorMode = promisify(this._processErrorIteratorModeCallback);
+
+  /** @internal */
+  private _processErrorIteratorModeCallback(error: AnyError, callback: Callback) {
```

Comment:
Can we just make this async?

Reply:
Ah I see, we don't want to have to do that everytime, but we should leave a comment saying thats why we did this, so we can removing it when serverSelection is awaitable

Reply:
We could also expose a promisified server selection function on the topology that we could make use of now throughout the driver.  Then we wouldn't have to promisify it each time.

And fwiw, I actually don't think it's bad to have the promisifying here.  It's not a hot code path - it's the error handling function.  So, this would either get called in one-off transient failures before the change stream resumes or twice in a row in the case of an resumable error that can't resume, and then the change stream would be closed.

thoughts?